core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,203 @@
+use zksync_queued_job_processor::JobProcessor;
+use zksync_types::witness_block_state::WitnessBlockState;
+use zksync_types::{L1BatchNumber, L2ChainId};
+
+use async_trait::async_trait;
+use tokio::task::JoinHandle;
+
+mod vm_interactions;
+
+use vm_interactions::{create_vm, execute_tx, get_miniblock_transition_state};
+
+#[derive(Debug)]
+pub struct BasicWitnessInputProducerJob {
+    l1_batch_number: L1BatchNumber,
+}
+
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        state_keeper_config: &StateKeeperConfig,
+        network_config: &NetworkConfig,
+    ) -> anyhow::Result<Self> {
+        // BEWARE, HERE BE DRAGONS.
+        // This usage here is a race condition waiting to happen.
+        // Context -- The VM run in BasicWitnessInputProducer MUST be identical to run from StateKeeper.
+        // Race condition scenario:
+        // 1. StateKeeper runs with a config A, which has value a for `validation_computational_gas_limit`
+        // 2. StateKeeper runs the batch and saves data to database.
+        // 3. Deployment is made and changes to config B, with value b for `validation_computational_gas_limit`.
+        // 4. BasicWitnessInputProducer loads config B
+        // 5. BasicWitnessInputProducer runs the L1Batch that was saved with config A using config B.
+        // In this scenario, the outputs will be different, which will make Witness Generation crash.
+        // This has been discussed between @evl and @deniallugo as a terrible solution, but best so far.
+        let validation_computational_gas_limit =
+            state_keeper_config.validation_computational_gas_limit;
+        // This variable is fine to be loaded from config.
+        // Today, this config changes only during regenesis.
+        // In the hyper-chains world, this code will be refactored (using the DB saved value instead of config)
+        let l2_chain_id = network_config.zksync_network_id;
+
+        Ok(BasicWitnessInputProducer {
+            connection_pool,
+            validation_computational_gas_limit,
+            object_store: store_factory.create_store().await.into(),
+            l2_chain_id,
+        })
+    }
+
+    async fn process_job_impl(
+        job: BasicWitnessInputProducerJob,
+        started_at: Instant,
+        connection_pool: ConnectionPool,
+        validation_computational_gas_limit: u32,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<WitnessBlockState> {
+        // This spawn_blocking is needed given PostgresStorage's interface is a mix of blocking and non-blocking
+        // The interface may be either refactored or broken down in a async and sync interface.
+        tokio::task::spawn_blocking(move || {
```

Comment:
`tokio::task::spawn_blocking` are primarily suitable for long-running computations or blocking I/O.
For other jobs witness-gen/circuit-synth it fine as they are long running cpu-intensive, I am assuming the computation performed here are also long running?

Reply:
`spawn_blocking` is mandatory here, and it's addressed in one of my comments.